File: lib_openweather/src/main/java/guepardoapps/library/openweather/common/tools/Logger.java
package guepardoapps.library.openweather.common.tools;

import android.support.annotation.NonNull;
import android.util.Log;

import java.io.Serializable;

public class Logger implements Serializable {

    private String _tag;
    private boolean _debuggingEnabled;

    public Logger(
            @NonNull String tag,
            boolean debuggingEnabled) {
        _tag = tag;
        _debuggingEnabled = debuggingEnabled;
        Debug("Created logger for " + _tag + "! Is Enabled: " + String.valueOf(_debuggingEnabled));
    }

    public Logger(@NonNull String tag) {
        this(tag, true);
    }

    public void SetEnable(boolean debuggingEnabled) {
        _debuggingEnabled = debuggingEnabled;
    }

    public void Verbose(@NonNull String message) {
        if (_debuggingEnabled) {
            if (message.length() > 0) {
                Log.v(_tag, message);
            }
        }
    }

    public void Debug(@NonNull String message) {
        if (_debuggingEnabled) {
            if (message.length() > 0) {
                Log.d(_tag, message);
            }
        }
    }

    public void Info(@NonNull String message) {
        if (_debuggingEnabled) {
            if (message.length() > 0) {
                Log.i(_tag, message);
            }
        }
    }

    public void Warn(@NonNull String message) {
        if (_debuggingEnabled) {
            if (message.length() > 0) {
                Log.w(_tag, message);
            }
        }
    }

    public void Error(@NonNull String message) {
        if (_debuggingEnabled) {
            if (message.length() > 0) {
                Log.e(_tag, message);
            }
        }
    }

    public void LogTest(
            @NonNull String message,
            boolean success) {
        if (_debuggingEnabled) {
            if (success) {
                if (message.length() > 0) {
                    Log.d(_tag, "TestResult:");
                    Log.i(_tag, message);
                }
            } else {
                if (message.length() > 0) {
                    Log.d(_tag, "TestResult:");
                    Log.e(_tag, message);
                }
            }
        }
    }
}

File: lib/src/main/java/guepardoapps/lib/openweather/utils/Logger.kt
package guepardoapps.lib.openweather.utils

import android.content.Context
import android.support.annotation.NonNull
import android.util.Log
import guepardoapps.lib.openweather.database.logging.DbHandler
import guepardoapps.lib.openweather.database.logging.DbLog
import guepardoapps.lib.openweather.database.logging.Severity
import java.sql.Date
import java.util.*

class Logger private constructor() {
    private var _loggingEnabled: Boolean = true
    private var _writeToDatabaseEnabled: Boolean = true

    private var _dbHandler: DbHandler? = null

    init {
    }

    private object Holder {
        val instance: Logger = Logger()
    }

    companion object {
        val instance: Logger by lazy { Holder.instance }
    }

    fun initialize(context: Context) {
        if (_dbHandler != null) {
            return
        }
        _dbHandler = DbHandler(context, null)
    }

    fun setLoggingEnabled(loggingEnabled: Boolean) {
        _loggingEnabled = loggingEnabled
    }

    fun isLoggingEnabled(): Boolean {
        return _loggingEnabled
    }

    fun setWriteToDatabaseEnabled(writeToDatabaseEnabled: Boolean) {
        _writeToDatabaseEnabled = writeToDatabaseEnabled
    }

    fun isWriteToDatabaseEnabled(): Boolean {
        return _writeToDatabaseEnabled
    }

    fun <T> verbose(@NonNull tag: String, @NonNull description: T) {
        if (_loggingEnabled) {
            Log.v(tag, description.toString())
            tryToWriteToDatabase(tag, description, Severity.Verbose)
        }
    }

    fun <T> debug(@NonNull tag: String, @NonNull description: T) {
        if (_loggingEnabled) {
            Log.d(tag, description.toString())
            tryToWriteToDatabase(tag, description, Severity.Debug)
        }
    }

    fun <T> info(@NonNull tag: String, @NonNull description: T) {
        if (_loggingEnabled) {
            Log.i(tag, description.toString())
            tryToWriteToDatabase(tag, description, Severity.Info)
        }
    }

    fun <T> warning(@NonNull tag: String, @NonNull description: T) {
        if (_loggingEnabled) {
            Log.w(tag, description.toString())
            tryToWriteToDatabase(tag, description, Severity.Warning)
        }
    }

    fun <T> error(@NonNull tag: String, @NonNull description: T) {
        if (_loggingEnabled) {
            Log.e(tag, description.toString())
            tryToWriteToDatabase(tag, description, Severity.Error)
        }
    }

    private fun <T> tryToWriteToDatabase(
            @NonNull tag:
            String, @NonNull description: T,
            severity: Severity) {
        if (_writeToDatabaseEnabled && _dbHandler != null) {
            _dbHandler?.addLog(
                    DbLog(-1,
                            Date(Calendar.getInstance().timeInMillis),
                            severity,
                            tag,
                            description.toString()))
        }
    }
}